### DISCUSSION TITLE
From 2.2.19 to 3.0.0-alpha-2 `tailwindExtractor` for PurgeCSS is missing

### DISCUSSION THREAD

**Discussion by frederikhors** [tailwindlabs/tailwindcss#6162] [Help]
I have a project in which I'm using Tailwind along with others CSS frameworks (like `animate.css`).

After the build I'm using PurgeCSS to remove what I don't need (there are approximately 40 KB only for `animate.css`). 

Coming to Tailwind v3 from 2.2.19 I have a problem with this code I was using:

```js
import { PurgeCSS, defaultOptions } from "purgecss";
import { tailwindExtractor } from "tailwindcss/lib/lib/purgeUnusedStyles";

const purgeCSSOptions = {
	content: ["./build/app/**/*.js", "...others..."],
	css: ["./build/app/**/*.css"],
	defaultExtractor: (content) => [
		...defaultOptions.defaultExtractor(content),
		...tailwindExtractor(content),
	]
};

// PurgeCSS()...
```

It cannot find `tailwindExtractor` anymore now with v3.

I tried to copy this code from `expandTailwindAtRules.js`:

```js
const PATTERNS = [
	/([^<>"'`\s]*\[\w*'[^"`\s]*'?\])/.source,
	/([^<>"'`\s]*\[\w*"[^"`\s]*"?\])/.source,
	/([^<>"'`\s]*\[\w*\('[^"'`\s]*'\)\])/.source,
	/([^<>"'`\s]*\[\w*\("[^"'`\s]*"\)\])/.source,
	/([^<>"'`\s]*\[\w*\('[^"`\s]*'\)\])/.source,
	/([^<>"'`\s]*\[\w*\("[^'`\s]*"\)\])/.source,
	/([^<>"'`\s]*\['[^"'`\s]*'\])/.source,
	/([^<>"'`\s]*\["[^"'`\s]*"\])/.source,
	/([^<>"'`\s]*\[[^"'`\s]+\][^<>"'`\s]*)/.source,
	/([^<>"'`\s]*[^"'`\s:])/.source,
].join("|");
const BROAD_MATCH_GLOBAL_REGEXP = new RegExp(PATTERNS, "g");
const INNER_MATCH_GLOBAL_REGEXP =
	/[^<>"'`\s.(){}[\]#=%]*[^<>"'`\s.(){}[\]#=%:]/g;
const builtInExtractors = {
	DEFAULT: (content) => {
		let broadMatches = content.match(BROAD_MATCH_GLOBAL_REGEXP) || [];
		let innerMatches = content.match(INNER_MATCH_GLOBAL_REGEXP) || [];
		return [...broadMatches, ...innerMatches];
	},
};
```

using it like this:

```js
const purgeCSSOptions = {
	content: ["./build/app/**/*.js", "...others..."],
	css: ["./build/app/**/*.css"],
	defaultExtractor: (content) => [
		...defaultOptions.defaultExtractor(content),
		...builtInExtractors.DEFAULT(content),
	]
};
```

but with this some classes does not work in final CSS bundle, example:

```
.active\:scale-95, .active\:scale-100, .after\:-translate-x-1\/2:after, .scale-100, .transform {
    --tw-translate-x: 0;
    --tw-translate-y: 0;
    --tw-rotate: 0;
    --tw-skew-x: 0;
    --tw-skew-y: 0;
    --tw-scale-x: 1;
    --tw-scale-y: 1;
    --tw-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
```

each `var(--tw-...)` value is undefined: `var(--tw-translate-x)`, `var(--tw-translate-y)`, `var(--tw-rotate)`, `var(--tw-skew-x)`, `var(--tw-skew-y)` and so on.

Using extractor copied from Tailwind internals changes code from this:

```css
*,
:after,
:before {
	border: 0 solid;
	box-sizing: border-box;
}
:after,
:before {
	--tw-content: "";
}
```

to this

```css
* {
	border: 0 solid;
	box-sizing: border-box;
}
```

which is totally wrong (especially for Tailwind 3).

![image](https://user-images.githubusercontent.com/41120635/142771407-fb57aff2-37e7-487b-b610-7e79b6d6e950.png)

How to fix this?

**Comment by frederikhors** [comment]
I think this is should be an issue, @adamwathan.

**Reply by adamwathan** [reply]
There’s no bug here so it’s not an issue, you are just asking for help about setting up PurgeCSS. The file you are referencing here has always been private and never been public API.

**Reply by frederikhors** [reply]
Ok. I got the wording wrong then.

But I think it's an issue that a used framework like yours doesn't have an official PurgeCSS extractor. This fact stops us upgrading to v3 (and we are also stuck with v2 due to the problem of var (- tw -...) undefined (even without PurgeCSS)).

Can I ask you if you are working on it or there is no effort on this problem?

**Reply by shmshd** [reply]
@frederikhors Why do you manually set PurgeCSS configuration when TailwindCSS does that all under the hood? Is there any specific reason to do so? If not, please go through this [docs on setting up PurgeCSS](https://tailwindcss.com/docs/optimizing-for-production#purge-css-options).
Remember, in v3, there is a slight change in purge config. Read [v3.0.0-alpha.1 changelog](https://github.com/tailwindlabs/tailwindcss/releases/tag/v3.0.0-alpha.1).

**Reply by frederikhors** [reply]
@shmshd I need a Tailwind extractor because I don't use Tailwind only in my project, there are other libraries that create a huge final bundle. Is it clear now?

**Reply by NovusOpus** [reply]
@frederikhors There must be better ways to achieve what you're trying to do here. For example you can configure your setup of **PurgeCSS** to only apply to those other libraries, without messing with tailwind files. Since tailwind already use their own implementation of **PurgeCSS** under the hood.

**Reply by frederikhors** [reply]
@NovusOpus Yes. And in fact I try to do this with the Tailwind extractor that should be there.

**Reply by NovusOpus** [reply]
But as @adamwathan mentioned, that file is not intended to be implemented as a public api, That means for use outside of tailwind. you can try, as there's some people in past discussions that tried to do something like what you're doing, but since it's not intended to be of public use, can't be considered an Issue or bug, since you're trying to use it for purposes outside of it's intended use.

**Reply by frederikhors** [reply]
@NovusOpus In fact I'm not asking for that file anymore. I'm asking for a new Tailwind 3 PurgeCSS extractor because I think that sooner or later we will all need it.

**Reply by NovusOpus** [reply]
@frederikhors In Tailwind CSS V3.0 there's no need of PurgeCSS for deleting unused classes, Since all the code is generated on the fly, Since it's going to use the [JIT Engine](https://tailwindcss.com/docs/just-in-time-mode) by default. That's why purge option is being replaced by content in tailwindconfig. Since it's not purging code anymore, but generating it  on-demand.

**Reply by frederikhors** [reply]
@NovusOpus I get it. We also said it above.

I have a huge bundle (20 graphics libraries) that weighs 5MB.

Using PurgeCSS it becomes a few KB.

All beautiful until I discovered that the Tailwind extractor I was using is no longer good for version 3.

Now I'm asking the world for an extractor for version 3.

I don't think it's heresy.

I think we will all need it.

Apart from the sacrosanct and lovable JIT mode.

**Reply by NovusOpus** [reply]
@frederikhors  You can still use PurgeCSS tho, just create a new discussion and ask for help of how to configure it. Probably you're going to get better help in the [PurgeCSS GitHub page](https://github.com/FullHuman/purgecss/discussions). but you can posted here as well, maybe someone using a setup like you can guide you in the right way.

Share what Development environment setup you're using, if you using Next, Nuxt, Gatsbty, Vite or others, and the path to the files you want to apply the purgeCSS.

**Reply by adamwathan** [reply]
I would recommend using PurgeCSS's control comments to not purge anything Tailwind generates (since it is effectively already purged), and then setup your extractor based on what you need to match your own class names:

```css
/* purgecss start ignore */
@tailwind base;
/* purgecss end ignore */

/* purgecss start ignore */
@tailwind components;
/* purgecss end ignore */

/* purgecss start ignore */
@tailwind utilities;
/* purgecss end ignore */

/* purgecss start ignore */
@tailwind variants;
/* purgecss end ignore */
```

This will likely be better for performance too since PurgeCSS won't have to scan those blocks 👍🏻 

I do see value in providing an extractor for people who want to use PurgeCSS in addition to what we're doing already for you, but unfortunately not something we are going to be able to prioritize in the near future.

You can craft a custom extractor based on our regular expressions here if it's helpful:

https://github.com/tailwindlabs/tailwindcss/blob/master/src/lib/expandTailwindAtRules.js#L9-L31

But that file is subject to refactoring and stuff, so don't rely on it directly.

At the end of the day again I definitely think it would be helpful for us to expose something here, but just can't promise a timeline on it or if it will ever actually become the highest priority item in our backlog and actually get worked on.

**Reply by frederikhors** [reply]
Now I recognize you @adamwathan! Thanks for the tip! I try and update everyone here.

**Reply by frederikhors** [reply]
There is a problem with comments and Tailwind with postcss: https://github.com/tailwindlabs/tailwindcss/issues/6182.

**Reply by danielehrhardt** [reply]
@frederikhors did you found a Solution?

**Reply by frederikhors** [reply]
@danielehrhardt I'm waiting for Vite 2.8 (which contains esbuild 0.14) and SvelteKit to update it. Then it will work with comments correctly.

**Reply by IHIutch** [reply]
Wanted to follow up here because I was experiencing an issue related to this. My solution ended up being to split out my "tailwind" file and everything else into 2 separate css files.

My postcss script looks like this `"dev:postcss": "postcss styles/*.scss -m -w -d app/styles --ext css"` and I have two separate files `hamburgers.scss` and `main.scss`.

Inside my `postcss.config.js` files I am doing something like this:
```
const handlePurge = (ctx) => {
  return ctx.file.basename !== 'main.scss'
    ? {
        '@fullhuman/postcss-purgecss': {
          content: ['./app/**/*.{js,ts,jsx,tsx}'],
      }
    : {}
}

module.exports = (ctx) => ({
  ...
  parser: 'postcss-scss', // this is only needed for scss files
  plugins: {
    ...handlePurge(ctx),
    ...other plugins...
  },
})
```
Which only applies `purgecss` to files that I specify. In my case, I have a file (`hamburger.scss`) that needs to be a `.scss` file because it uses SASS variables, but the rest of my css is located in`main.scss`. This allows me to purge outputted `hamburger.css` file separately from my other css file, where I've included Tailwind and other styles that can use Tailwind's default compiler as expected.

You end up with at least two files, but for me it works without anything that feels like a hack.
